File: sdk/samples/02_qos/python/latency_budget.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0 OR MIT
# Copyright (c) 2025-2026 naskel.com

"""
HDDS Sample: Latency Budget (Python)

Demonstrates LATENCY_BUDGET QoS for delivery latency hints.
A low budget (0ms) requests immediate delivery, while a higher budget
(100ms) allows the middleware to batch or optimize delivery.

Usage:
    python latency_budget.py        # Subscriber (monitors both topics)
    python latency_budget.py pub    # Publisher (sends on both topics)
"""

import sys
import time

sys.path.insert(0, str(__file__ + "/../../01_basics/python"))

import hdds
from generated.HelloWorld import HelloWorld

LOW_LATENCY_MS = 0
HIGH_LATENCY_MS = 100
NUM_MESSAGES = 5


def run_publisher(participant):
    """Publish to two topics with different latency budgets."""
    qos_low = hdds.QoS.reliable().latency_budget_ms(LOW_LATENCY_MS)
    qos_high = hdds.QoS.reliable().latency_budget_ms(HIGH_LATENCY_MS)

    writer_low = participant.create_writer("LowLatencyTopic", qos=qos_low)
    writer_high = participant.create_writer("BatchedTopic", qos=qos_high)

    print(f"Publishing {NUM_MESSAGES} messages on each topic...")
    print(f"  LowLatencyTopic: latency_budget = {LOW_LATENCY_MS}ms (immediate)")
    print(f"  BatchedTopic:    latency_budget = {HIGH_LATENCY_MS}ms (batched)\n")

    start = time.monotonic()

    for i in range(NUM_MESSAGES):
        elapsed = int((time.monotonic() - start) * 1000)

        msg_low = HelloWorld(id=i + 1, message=f"Low-latency #{i + 1}")
        writer_low.write(msg_low.serialize())
        print(f"  [{elapsed:5d}ms] Sent LowLatency  id={msg_low.id}")

        msg_high = HelloWorld(id=i + 1, message=f"Batched #{i + 1}")
        writer_high.write(msg_high.serialize())
        print(f"  [{elapsed:5d}ms] Sent Batched     id={msg_high.id}")

        time.sleep(0.3)

    print("\nDone publishing on both topics.")


def run_subscriber(participant):
    """Monitor both topics and compare delivery timing."""
    qos_low = hdds.QoS.reliable().latency_budget_ms(LOW_LATENCY_MS)
    qos_high = hdds.QoS.reliable().latency_budget_ms(HIGH_LATENCY_MS)

    reader_low = participant.create_reader("LowLatencyTopic", qos=qos_low)
    reader_high = participant.create_reader("BatchedTopic", qos=qos_high)

    waitset = hdds.WaitSet()
    waitset.attach(reader_low)
    waitset.attach(reader_high)

    print("Subscribing to both topics...")
    print(f"  LowLatencyTopic: latency_budget = {LOW_LATENCY_MS}ms")
    print(f"  BatchedTopic:    latency_budget = {HIGH_LATENCY_MS}ms\n")

    received_low = 0
    received_high = 0
    start = time.monotonic()
    timeouts = 0

    while timeouts < 3:
        if waitset.wait(timeout_secs=2.0):
            while True:
                data = reader_low.take()
                if data is None:
                    break
                msg = HelloWorld.deserialize(data)
                elapsed = int((time.monotonic() - start) * 1000)
                print(f"  [{elapsed:5d}ms] LowLatency  received id={msg.id}")
                received_low += 1

            while True:
                data = reader_high.take()
                if data is None:
                    break
                msg = HelloWorld.deserialize(data)
                elapsed = int((time.monotonic() - start) * 1000)
                print(f"  [{elapsed:5d}ms] Batched     received id={msg.id}")
                received_high += 1

            timeouts = 0
        else:
            timeouts += 1

    print()
    print("-" * 60)
    print(f"Summary:")
    print(f"  LowLatency (budget={LOW_LATENCY_MS}ms): {received_low} messages")
    print(f"  Batched    (budget={HIGH_LATENCY_MS}ms): {received_high} messages")
    print()
    print("Note: Latency budget is a HINT to the middleware.")
    print("Actual delivery depends on transport and implementation.")
    print("-" * 60)


def main():
    is_publisher = len(sys.argv) > 1 and sys.argv[1] == "pub"

    hdds.logging.init(hdds.LogLevel.INFO)

    print("=" * 60)
    print("Latency Budget Demo")
    print("QoS: LATENCY_BUDGET - delivery latency hints for optimization")
    print("=" * 60)

    with hdds.Participant("LatencyBudgetDemo") as participant:
        if is_publisher:
            run_publisher(participant)
        else:
            run_subscriber(participant)


if __name__ == "__main__":
    main()

File: sdk/samples/02_qos/rust/src/bin/latency_budget.rs
// SPDX-License-Identifier: Apache-2.0 OR MIT
// Copyright (c) 2025-2026 naskel.com

//! # HDDS Sample: Latency Budget QoS
//!
//! Demonstrates **LATENCY_BUDGET** QoS - provides delivery latency hints
//! to the middleware, allowing it to optimize batching and scheduling.
//!
//! ## How LATENCY_BUDGET Works
//!
//! ```text
//! Budget = 0ms (immediate delivery):
//!
//!   Writer:  [msg]----->  Middleware  ----->  Reader
//!                         (send now)         (instant)
//!
//! Budget = 100ms (batching allowed):
//!
//!   Writer:  [msg1]--+
//!            [msg2]--+--> Middleware  ------> Reader
//!            [msg3]--+   (batch up            (receives batch
//!                         to 100ms)            within 100ms)
//! ```
//!
//! ## Use Cases
//!
//! - **Real-time control vs analytics**: Control loops need budget=0,
//!   analytics can tolerate batched delivery
//! - **Trading vs reporting**: Order execution needs minimal latency,
//!   position reports can be batched
//! - **Bandwidth optimization**: Higher budgets allow more efficient
//!   network utilization through batching
//!
//! ## Running the Sample
//!
//! ```bash
//! # Single-process mode (recommended):
//! cargo run --bin latency_budget
//!
//! # Two-terminal mode:
//! # Terminal 1 - Subscriber
//! cargo run --bin latency_budget -- sub
//!
//! # Terminal 2 - Publisher
//! cargo run --bin latency_budget -- pub
//! ```

use std::env;
use std::sync::Arc;
use std::thread;
use std::time::{Duration, Instant};

#[allow(dead_code)]
mod generated {
    include!("../../../../01_basics/rust/generated/hello_world.rs");
}

use generated::hdds_samples::HelloWorld;

/// Number of messages per topic
const NUM_MESSAGES: u32 = 5;
/// Publish interval in milliseconds
const PUBLISH_INTERVAL_MS: u64 = 200;

// =============================================================================
// Publisher
// =============================================================================

fn run_publisher(participant: &Arc<hdds::Participant>) -> Result<(), hdds::Error> {
    // -------------------------------------------------------------------------
    // Low-Latency Writer (budget = 0ms)
    // -------------------------------------------------------------------------
    //
    // Middleware should deliver immediately, no batching allowed.

    let low_latency_qos = hdds::QoS::reliable().latency_budget_millis(0);

    let low_writer = participant.create_writer::<HelloWorld>("LowLatencyTopic", low_latency_qos)?;

    // -------------------------------------------------------------------------
    // Batched Writer (budget = 100ms)
    // -------------------------------------------------------------------------
    //
    // Middleware may batch messages within the 100ms window for efficiency.

    let batched_qos = hdds::QoS::reliable().latency_budget_millis(100);

    let batch_writer = participant.create_writer::<HelloWorld>("BatchedTopic", batched_qos)?;

    println!(
        "Publishing {} messages on each topic at {}ms intervals...\n",
        NUM_MESSAGES, PUBLISH_INTERVAL_MS
    );

    let start = Instant::now();

    for i in 0..NUM_MESSAGES {
        let elapsed = start.elapsed().as_millis();

        // Send on low-latency topic
        let low_msg = HelloWorld::new(format!("LowLatency #{}", i + 1), i + 1);
        low_writer.write(&low_msg)?;
        println!(
            "  [{:5}ms] Sent LowLatency  #{} (budget=0ms)",
            elapsed,
            i + 1
        );

        // Send on batched topic
        let batch_msg = HelloWorld::new(format!("Batched #{}", i + 1), i + 1);
        batch_writer.write(&batch_msg)?;
        println!(
            "  [{:5}ms] Sent Batched     #{} (budget=100ms)",
            elapsed,
            i + 1
        );

        thread::sleep(Duration::from_millis(PUBLISH_INTERVAL_MS));
    }

    println!("\nPublisher finished.");
    Ok(())
}

// =============================================================================
// Subscriber
// =============================================================================

fn run_subscriber(participant: &Arc<hdds::Participant>) -> Result<(), hdds::Error> {
    // -------------------------------------------------------------------------
    // Low-Latency Reader
    // -------------------------------------------------------------------------

    let low_latency_qos = hdds::QoS::reliable().latency_budget_millis(0);

    let low_reader = participant.create_reader::<HelloWorld>("LowLatencyTopic", low_latency_qos)?;

    // -------------------------------------------------------------------------
    // Batched Reader
    // -------------------------------------------------------------------------

    let batched_qos = hdds::QoS::reliable().latency_budget_millis(100);

    let batch_reader = participant.create_reader::<HelloWorld>("BatchedTopic", batched_qos)?;

    let waitset = hdds::dds::WaitSet::new();
    waitset.attach_condition(low_reader.get_status_condition())?;
    waitset.attach_condition(batch_reader.get_status_condition())?;

    println!("Listening on both topics...\n");

    let mut low_received = 0u32;
    let mut batch_received = 0u32;
    let mut low_total_latency_us: u128 = 0;
    let mut batch_total_latency_us: u128 = 0;
    let start = Instant::now();
    let mut timeouts = 0;

    let target = NUM_MESSAGES * 2;

    while (low_received + batch_received) < target && timeouts < 3 {
        match waitset.wait(Some(Duration::from_secs(2))) {
            Ok(triggered) if !triggered.is_empty() => {
                let recv_time = start.elapsed();

                // Check low-latency reader
                while let Some(msg) = low_reader.take().ok().flatten() {
                    let elapsed = recv_time.as_millis();
                    println!("  [{:5}ms] Received LowLatency  #{}", elapsed, msg.count);
                    low_received += 1;
                    low_total_latency_us += recv_time.as_micros();
                }

                // Check batched reader
                while let Some(msg) = batch_reader.take().ok().flatten() {
                    let elapsed = recv_time.as_millis();
                    println!("  [{:5}ms] Received Batched     #{}", elapsed, msg.count);
                    batch_received += 1;
                    batch_total_latency_us += recv_time.as_micros();
                }

                timeouts = 0;
            }
            Ok(_) | Err(hdds::Error::WouldBlock) => {
                timeouts += 1;
            }
            Err(e) => eprintln!("Wait error: {:?}", e),
        }
    }

    // -------------------------------------------------------------------------
    // Summary
    // -------------------------------------------------------------------------

    println!("\n{}", "-".repeat(50));
    println!(
        "Low-Latency (budget=0ms):   {} messages received",
        low_received
    );
    if low_received > 0 {
        let avg = low_total_latency_us / low_received as u128;
        println!("Low-Latency avg latency:  {} us", avg);
    }
    println!(
        "Batched     (budget=100ms): {} messages received",
        batch_received
    );
    if batch_received > 0 {
        let avg = batch_total_latency_us / batch_received as u128;
        println!("Batched avg latency:      {} us", avg);
    }
    println!("{}", "-".repeat(50));
    println!("Note: Actual latency differences depend on middleware");
    println!("implementation and network conditions.");
    println!("{}", "-".repeat(50));

    Ok(())
}

// =============================================================================
// Single-Process Demo
// =============================================================================

fn run_single_process(participant: &Arc<hdds::Participant>) -> Result<(), hdds::Error> {
    // -------------------------------------------------------------------------
    // Setup writers and readers for both latency profiles
    // -------------------------------------------------------------------------

    let low_latency_qos = hdds::QoS::reliable().latency_budget_millis(0);

    let batched_qos = hdds::QoS::reliable().latency_budget_millis(100);

    let low_writer =
        participant.create_writer::<HelloWorld>("LowLatencyTopic", low_latency_qos.clone())?;
    let batch_writer =
        participant.create_writer::<HelloWorld>("BatchedTopic", batched_qos.clone())?;

    let low_reader = participant.create_reader::<HelloWorld>("LowLatencyTopic", low_latency_qos)?;
    let batch_reader = participant.create_reader::<HelloWorld>("BatchedTopic", batched_qos)?;

    let waitset = hdds::dds::WaitSet::new();
    waitset.attach_condition(low_reader.get_status_condition())?;
    waitset.attach_condition(batch_reader.get_status_condition())?;

    // Allow time for discovery
    thread::sleep(Duration::from_millis(100));

    println!("Sending {} messages on each topic...\n", NUM_MESSAGES);

    let start = Instant::now();
    let mut low_arrivals: Vec<u128> = Vec::new();
    let mut batch_arrivals: Vec<u128> = Vec::new();

    for i in 0..NUM_MESSAGES {
        let send_time = start.elapsed().as_millis();

        // Publish on both topics simultaneously
        let low_msg = HelloWorld::new(format!("LowLatency #{}", i + 1), i + 1);
        low_writer.write(&low_msg)?;

        let batch_msg = HelloWorld::new(format!("Batched #{}", i + 1), i + 1);
        batch_writer.write(&batch_msg)?;

        println!("  [{:5}ms] Sent #{} on both topics", send_time, i + 1);

        // Brief wait to allow delivery
        thread::sleep(Duration::from_millis(10));

        // Check for arrivals
        match waitset.wait(Some(Duration::from_millis(200))) {
            Ok(triggered) if !triggered.is_empty() => {
                while let Some(_msg) = low_reader.take().ok().flatten() {
                    let arrival = start.elapsed().as_millis();
                    low_arrivals.push(arrival);
                    println!(
                        "  [{:5}ms]   -> LowLatency arrived (delta={}ms)",
                        arrival,
                        arrival - send_time
                    );
                }
                while let Some(_msg) = batch_reader.take().ok().flatten() {
                    let arrival = start.elapsed().as_millis();
                    batch_arrivals.push(arrival);
                    println!(
                        "  [{:5}ms]   -> Batched arrived (delta={}ms)",
                        arrival,
                        arrival - send_time
                    );
                }
            }
            Ok(_) | Err(hdds::Error::WouldBlock) => {}
            Err(e) => eprintln!("Wait error: {:?}", e),
        }

        thread::sleep(Duration::from_millis(PUBLISH_INTERVAL_MS));
    }

    // Drain any remaining messages
    thread::sleep(Duration::from_millis(200));
    if let Ok(triggered) = waitset.wait(Some(Duration::from_millis(500))) {
        if !triggered.is_empty() {
            while let Some(_msg) = low_reader.take().ok().flatten() {
                low_arrivals.push(start.elapsed().as_millis());
            }
            while let Some(_msg) = batch_reader.take().ok().flatten() {
                batch_arrivals.push(start.elapsed().as_millis());
            }
        }
    }

    // -------------------------------------------------------------------------
    // Summary
    // -------------------------------------------------------------------------

    println!("\n{}", "-".repeat(50));
    println!(
        "LowLatency (budget=0ms):   {} of {} delivered",
        low_arrivals.len(),
        NUM_MESSAGES
    );
    println!(
        "Batched    (budget=100ms): {} of {} delivered",
        batch_arrivals.len(),
        NUM_MESSAGES
    );
    println!("{}", "-".repeat(50));
    println!("Note: With budget=0ms the middleware attempts immediate");
    println!("delivery. With budget=100ms, batching may occur.");
    println!("Actual behavior depends on the transport layer.");
    println!("{}", "-".repeat(50));

    Ok(())
}

// =============================================================================
// Main
// =============================================================================

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    let mode = args.get(1).map(|s| s.as_str());

    println!("{}", "=".repeat(60));
    println!("HDDS Latency Budget QoS Sample");
    println!("Delivery latency hints for middleware optimization");
    println!("{}\n", "=".repeat(60));

    let participant = hdds::Participant::builder("LatencyBudgetDemo")
        .with_transport(hdds::TransportMode::IntraProcess)
        .build()?;

    match mode {
        Some("pub") => run_publisher(&participant)?,
        Some("sub") => run_subscriber(&participant)?,
        _ => run_single_process(&participant)?,
    }

    Ok(())
}
